### DISCUSSION TITLE
How to always launch certain apps on certain workspaces?

### DISCUSSION THREAD

**Discussion by vadim-su** [pop-os/cosmic-epoch#1129] [Q&A]
And another small question: can I allow moving to any workspace if the previous one doesn't contain any windows?

**Comment by jacobgkau** [comment]
I don't understand what "relate" means. You can already drag windows between different workspaces, or move them with keyboard shortcuts.

Do you mean you want an app to always launch on a specific workspace?

**Reply by vadim-su** [reply]
Correct. I want to open Discord and Slack windows only on the fourth workspace. Is it possible?

**Reply by jacobgkau** [reply]
I've renamed your discussion's title to be more clear about your feature request.

**Reply by vadim-su** [reply]
Thanks

**Comment by nevdelap** [comment]
@vadim-su pop-os/cosmic-epoch#908 maybe covers what you're asking for?

**Reply by vadim-su** [reply]
Tying a specific workspace to specific applications is a matter of personal preference for me. I find it convenient to organize my work in this manner. I have implemented this approach in various desktop environments, including KDE and Hyprland.

**Reply by vadim-su** [reply]
Sorry, I lost part of the answer. It looks like it is what I need

**Comment by vadim-su** [comment]
In general, I prefer the next schema: 1st workspace for work, 2nd workspace for searching (ai, browser), 3rd player and messengers, 4th for discord and other. It's super comfortably to my mindset and it would be great if I reproduce the same behavior in cosmic de

**Comment by ctretyak** [comment]
I think you mean that we need configuration like i3wm, when you can assign app to certain ws number on open by specific rules (app class name, etc.), e.g.:
```
assign [class="Alacritty"] $ws2

assign [class="^Google-chrome$"] $ws3

assign [class="TelegramDesktop"] $ws5

assign [class="obsidian"] $ws6

assign [class="Studio 3T"] $ws8

assign [class="DBeaver"] $ws8

for_window [class="Spotify"] move scratchpad

assign [all class="^(?!xdg-desktop-portal-gnome)"] $ws7
```

**Reply by vadim-su** [reply]
Thank you. I will check the solution

**Comment by joelirish** [comment]
I wanted to do the very same thing, and I found that the 'auto-move-windows' extension in the [gnome-shell-extensions](https://gitlab.gnome.org/GNOME/gnome-shell-extensions#extensions) does just that.

In the extension's settings, you choose your given application, then indicate the workspace you want it live whenever it's running (1, 2, 3, et c.) and that's it. 

(Also, FWIW, I don't recall if I had to enable the extension myself, or if it already was by default.)

**Reply by justresisting** [reply]
I also wanted to do this and also found that `auto-move-windows` does exactly what I want. 

I also don't recall if I had to enable the extension but I did do experimenting and most likely did install it. 

It was absolutely simple and intuitive to set up my desired app-specific workspace conditions. It took less than five minutes to set up my desired workflows for all of the apps I wanted.

**Reply by briandipalma** [reply]
Gnome extensions work on cosmic epoch? I thought cosmic epoch only supported cosmic applets.

**Reply by justresisting** [reply]
You are right. 

I was using 22.02 when I answered this. There are so many repositories and I had searched quite a while to find this thread. I upgraded from 22.04 to 24.04 through the command line within the last two weeks and still have some gnome-shell-extensions installed and residual-configuration files leftover. 

I am going to do a complete reformat and reinstallation of the operating system today because I have run into a number of issues and I want to have a fresh start to make sure the way I upgraded isn't causing any of them.

**Comment by ArkieSoft** [comment]
A feature like this would be very nice. At least allow a command line set those.

**Comment by stigman9393** [comment]
yeah, this is what still keeps me in gnome. automove window to workspace. I hope it will appear also in cosmic. If this option will be added, i switch to cosmic. Its stable enough but i cant live without that option

**Comment by piromanrs** [comment]
This is a must have feature for me, but it also needs to be able to auto start apps after OS is started. Basically when I turn on my computer, I get specific apps on different workspaces and I use specific shortcuts to access those workspaces, so I don't need to search for a started app, I know in which workspace it is and how to access it with single keypress!

**Reply by briandipalma** [reply]
Autostart can be achieved with autostart script: https://github.com/briandipalma/iac/tree/main/dotfiles/brians-infinitybookpro14-popos/autostart and https://github.com/briandipalma/iac/blob/main/setup/links.sh#L24

**Reply by piromanrs** [reply]
Great, thank you very much!

**Comment by piromanrs** [comment]
We got Pop OS 24.04 stable today, so I'm asking again about this issue. @briandipalma solution is very old so it might not be the best one for Cosmic (if it is for Cosmic at all? Yup I am that stupid and have no idea if it's for gnome)

**Reply by briandipalma** [reply]
Autostart configuration is standard. I've moved to CachyOS and KDE Plasma and it supports this feature but the autostart part is achieved via KDE creating autostart config for you in the background when you use the KDE GUI to select applications to autostart. This is the autostart configuration KDE created for me for example https://github.com/briandipalma/iac/tree/main/dotfiles/brians-infinitybookpro14-cachyos/autostart

**Reply by piromanrs** [reply]
Thanks @briandipalma. So this will work on Cosmic?

**Reply by briandipalma** [reply]
It will autostart the apps but they won't be placed in a specific workspace, I had to do that manually in Cosmic. https://specifications.freedesktop.org/autostart/latest/

**Reply by piromanrs** [reply]
Thank you @briandipalma

**Comment by HaruNashii** [comment]
It’s very strange that this basic feature isn’t in Cosmic yet. How was it released without such an essential feature? It’s a must-have for many people, and I don’t understand how anyone can use workspaces without it.

I’m very excited to use Cosmic. The Cosmic team has done—and is doing—an incredible job creating a fast, beautiful, and secure DE. But without this feature, it turns what could be great into a big no for my workflow, and I’m sure I’m not the only one who feels this way.

**Comment by sanathusk** [comment]
I am on the latest 24.04 version, looking forward work this feature or guidance on how to achieve this, if it is already supported

**Comment by AryaTilde** [comment]
I started using COSMIC recently, and it immediately has become my favourite and main go to. I Loved Gnome but it had many, many many issues for me and my gaming. And KDE, it worked fine but I just never managed to feel comfortable in it. COSMIC has all I ever wanted. Functional, comfortable, adjustable; and oh how elegant and pleasing. I love it way too much.

This feature, to be able to assign something like discord, or a browser, always on a certain workspace, would really really bring the experience to an elevated level for me.

And I don't know if anyone else would love it as much as me, but if it was possible to have a certain, numbered or named workspace, activated with a certain keybinding, but NOT on the default wheel of workspaces you have to navigate through right and left; and on overview, have it at the end seperated with a | line, from the rest. I would jump up and down from happiness to have that one day. That would be just perfect. That one really would be comfort and functionality at it's ever peaking maximum.

**Comment by estin** [comment]
My current solution is [cos-cli](https://github.com/estin/cos-cli), which sends requests to the COSMIC compositor to move apps to specific workspaces.

Examples

Find app id

```console
cos-cli info
```

```
Apps:
        org.wezfurlong.wezterm (title: [Z] [3/3] ~/prj/cos-cli)
        org.telegram.desktop._d7a3f789405d28ee7edfc66f96ed074d (title: ВОРОНЕЖСКОЕ НЕБО)
        firefox (title: estin/cos-cli: A CLI utility for COSMIC Wayland toplevel and workspace management — Mozilla Firefox)
        AmneziaVPN (title: AmneziaVPN)
Workspaces:
        Workspace: 1
        Workspace: 2
        Workspace: 3
        Workspace: 4
 Outputs:
        Output: HDMI-A-1
```

Move it

````console
cos-cli move -a firefox -w 1
````

````console
Connected to cosmic toplevel manager!
Move firefox to 1
````

Tested on 

````console
❯ uname
 kernel-name      | Linux
 nodename         | archlinux
 kernel-release   | 6.18.3-arch1-1
 kernel-version   | #1 SMP PREEMPT_DYNAMIC Fri, 02 Jan 2026 17:52:55 +0000
 machine          | x86_64
 operating-system | GNU/Linux

❯ cosmic-comp --version
cosmic-comp 1.0.0 (git commit 8a5d78dbb0eef579bc6afab5315363e872a15185)
````

My usage of this cli - simple script to start up apps and move them to a workspaces

**Reply by lcherpit** [reply]
Thank you for this contribution! It works perfectly with single-instance applications! How do I identify (and move) multi-instance applications like the one below?

```
❯ cos-cli info
Apps:
	kitty (title: user@host:~/dir0/subdir)
	jetbrains-toolbox (title: Toolbox)
	firefox (title: website title 1 current active tab)
	firefox (title: website title 2 current active tab)
	firefox (title: website title 3 current active tab)
	...
	kitty (title: Home)
	kitty (title: user@host:~)
```

**Reply by estin** [reply]
@lcherpit feel free to open issues on the target repo. cos-cli is experimental and a personal project built for my own needs. It will be supported until the Cosmic team implements its own CLI/API.

I recently pushed the following changes:
- move multiple apps found by an app_id pattern.
- move a certain app by its index (the index in the list of discovered apps) for cases when there are many windows with the same app_id.

**Reply by beemdvp** [reply]
Just wanted to say THANK YOU for this. My setup is absolutely perfect and was the last puzzle I wanted from gnome, much appreciated for this tool!

**Reply by AryaTilde** [reply]
Very good work @estin !
Thank you very much!

I wanted to ask, is it possible to have the script to 
_1. if the app is open on say workspace 2, just switch screen to workspace 2
2. if the app is not open, then open it, and move it to workspace 2, and focus that workspace as well_
What I would love is 2 extra functions if possible: "_To be able to move the focus with a window when moving it_", and "_only focusing it when it's already open_".
I intend to use this with keybind on keyboard. For example to have librewolf binded to super+b (with the cos-cli script), and when pressed, have it either just go to workspace 2 if already open and there. And if not, move it there with focus, or launch it, move it there, and focus it.
It would be truly amazing to have it that way! If you can please see if it's possible to add this function, I would be very thankful!

Edit: I wanted to try and do this same thing by binding the script as is, to a button in cosmic, and also bind "Move to Workspace 2" to the same button, which would basically fulfill the same function I need. But it unfortunately doesn't let us bind multiple separate things to a single button in cosmic (Which would honestly not make sense other than this specific situation). Thus I'm asking you if possible to be able to do the workspace switching with cli command as well, so that it can be appended to the script, and added in the single custom command.

Edit 2: What I described above, making it 2 functions, might be overly complex. Only having cos-cli have a command to switch workspace with cli command, would fulfill the same function too.
If possible have a command like this to command switching to workspace 2:
`cos-cli switch 2`
And then that way, it can be ran before/after the main move command, and would do exactly what I want (Yes, librewolf would still launch a second instance, but single instance applications like discord don't do it and basically just switch to their workspace, and worst case with librewolf I'd just close the newly made one with my super+q; so if just having a "switch" function for cos-cli is possible, it would be truly amazing.)

Thank you again!

**Reply by estin** [reply]
@AryaTilde I don't find way to activate workspace, but activate certain app works now.

cos-cli now can
- print info in json format `cos-cli info --json` to simplify parse it and find index
- activate app command `cos-cli activate -i 0` 

Please check the cos-cli readme

Sorry for my poor English

**Comment by karlskewes** [comment]
+1. I used this feature every day with i3 and Hyprland and miss it with Cosmic. It's a great workflow.

I've raised a couple of PR's that work for me using file based configuration. Ready to adjust as required. 
See: https://github.com/pop-os/cosmic-comp/pull/2214

**Comment by dariusmiles2123** [comment]
Coming from Gnome and having an old computer to test PopOS, I fully agree with the fact that we need to be able to automatically move applications to a certain workspace easily.
